File: sentimental-credit/credit.py
# replicates credit.c

import re


def checksum(numberS):
    # Implements Luhn's algorithm
    number = int(numberS)
    sum = 0
    for i in range(1, len(numberS) + 1):
        digit = number % 10
        number = int(number / 10)

        # Checking every other digit starting from the second-to-last
        if i % 2 == 0:
            prod = digit * 2
            while prod != 0:
                digit2 = prod % 10
                prod = int(prod / 10)

                sum = sum + digit2
        else:
            sum = sum + digit

    # Conclusion of the algo
    if sum % 10 == 0:
        return True
    else:
        return False


def main():
    # Card number
    card = input("Number: ")

    # Check number of digits
    hint = ''

    # Checking if it looks like a credit card from any known flag
    if re.search("^((?:34|37)\d{13})$", card) is None:
        if re.search("^((?:51|52|53|54|55)\d{14})$", card) is None:
            if re.search("^((?:4)(?:\d{12}|\d{15}))$", card) is None:
                print("INVALID")
                return -1
            else:
                hint = "VISA"
        else:
            hint = "MASTERCARD"
    else:
        hint = "AMEX"

    if checksum(card):
        print(hint)
        return 0
    else:
        print("INVALID")
        return 1


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Getting input number
    long cardNumber = get_long("Number: ");

    // defining some variables
    int size = 0;
    string flag = "INVALID\n";

    // Checking number of digits
    long temp = cardNumber;
    while (temp != 0)
    {
        int digit = temp % 10;
        temp = temp / 10;
        size++;
    }

    // Checking flag
    temp = cardNumber;
    int lastDigit = 0;
    for (int b = 1; b <= size; b++)
    {
        int digit = temp % 10;
        temp = temp / 10;

        // if we're on the second to last digit
        if (b == size - 1)
        {
            // switch case for deciding on flag
            switch (temp % 10)
            {
                case 3 :
                    if (digit == 4 || digit == 7)
                    {
                        flag = "AMEX\n";
                    }
                    break;
                case 4 :
                    flag = "VISA\n";
                    break;
                case 5 :
                    if (digit >= 1 && digit <= 5)
                    {
                        flag = "MASTERCARD\n";
                    }
                    break;
            }
        }

    }


    if (size == 13 || size == 15 || size == 16)
    {
        int sum = 0;
        temp = cardNumber;


        // Luhn's algorithm
        for (int i = 1; i <= size; i++)
        {
            int digit = temp % 10;
            temp = temp / 10;

            // Checking every other digit starting from the second-to-last
            if (i % 2 == 0)
            {
                int prod = digit * 2;
                while (prod != 0)
                {
                    int digit2 = prod % 10;
                    prod = prod / 10;

                    sum = sum + digit2;
                }
            }
            else
            {
                sum = digit + sum;
            }
        }

        // Conclusion of the algo
        if (sum % 10 == 0)
        {
            printf("%s", flag);
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }


}